crates/iota-data-ingestion-core/README.md

```diff
@@ -0,0 +1,352 @@
+    // Provide a remote URL to download checkpoints
+    remote_store_url: Option<String>,
+    // Any config key-value pair for the Remote fetcher
+    remote_store_options: Vec<(String, String)>,
+    // Customize and optimize the Checkpoint fetch behavior
+    options: ReaderOptions,
+) -> (
+    Self,
+    // Channel the Workers will listen and process & store checkpoints
+    mpsc::Receiver<CheckpointData>,
+    // Channel on which workers will notify GC of the checkpoint
+    mpsc::Sender<CheckpointSequenceNumber>,
+    // Graceful shutdown
+    oneshot::Sender<()>,
+)
+```
+
+The `ReaderOptions` struct provides options for configuring how the checkpoint reader fetches new checkpoints.
+
+```rust
+pub struct ReaderOptions {
+    // How often to check for new checkpoints
+    // Lower values mean faster detection but more CPU usage
+    // Default: 100ms
+    pub tick_interval_ms: u64,
+    // Network request timeout
+    // Applies to remote store operations
+    // Default: 5 seconds
+    pub timeout_secs: u64,
+    // Maximum concurrent remote requests
+    // Higher values increase throughput but use more resources
+    // Default: 10
+    pub batch_size: usize,
+    // Memory limit for processing batch checkpoints
+    // 0 means no limit
+    // Helps prevent OOM issues
+    pub data_limit: usize,
+}
+```
+
+The `run()` method is responsible for starting the actor. It creates an [inotify](https://man7.org/linux/man-pages/man7/inotify.7.html) listener to monitor the local directory for new files. The method then enters a loop with a `tokio::select!` statement to handle three execution branches:
+
+```rust
+loop {
+    tokio::select! {
+        // Graceful shutdown
+        _ = &mut self.exit_receiver => break,
+        // Receive from Workers the processed checkpoint sequence numbers, this means that the
+        // files are no longer needed and a GC can be safely performed
+        Some(gc_checkpoint_number) = self.processed_receiver.recv() => {
+            self.gc_processed_files(gc_checkpoint_number).expect("Failed to clean the directory");
+        }
+        // `self.sync()` is invoked either after waiting `self.options.tick_interval_ms`
+        // without a new file notification, or immediately upon receiving such a notification.
+        Ok(Some(_)) | Err(_) = timeout(Duration::from_millis(self.options.tick_interval_ms), inotify_recv.recv())  => {
+            self.sync().await.expect("Failed to read checkpoint files");
+        }
+    }
+}
+```
+
+The `sync()` method manages the retrieval of checkpoints, coordinating between local and (optionally) remote sources. It prioritizes local checkpoints, using the `read_local_files()` method to fetch them from the local directory. Remote fetching, performed by the `remote_fetch()` method, is only activated if the local checkpoints are missing or lag behind the expected sequence. This design ensures that local storage is the primary source of truth.
```

Comment:
Here too links to the methods mentioned (`sync`, `read_local_files`, `remote_fetch`) would help the reader to go deeper into the implementation.

Reply:
[fixed](https://github.com/iotaledger/iota/pull/4570/commits/c58ed8cab8960cc11588dd32e7a540ae8ebe4c3c#diff-31a167158fe911c55387e5f89c28edd3fd006968b761a550e15148aee28192cbR118-R121)